strawberry/types/types.py

```diff
@@ -13,29 +15,237 @@
+
+        return type(self)(of_type_copy)
+
+    @property
+    def is_graphql_generic(self) -> bool:
+        from strawberry.schema.compat import is_graphql_generic
+
+        type_ = self.of_type
+
+        return is_graphql_generic(type_)
+
+    def has_generic(self, type_var: TypeVar) -> bool:
+        if isinstance(self.of_type, StrawberryType):
+            return self.of_type.has_generic(type_var)
+        return False
+
+
+class StrawberryList(StrawberryContainer): ...
+
+
+class StrawberryOptional(StrawberryContainer): ...
+
+
+class StrawberryTypeVar(StrawberryType):
+    def __init__(self, type_var: TypeVar) -> None:
+        self.type_var = type_var
+
+    def copy_with(
+        self, type_var_map: Mapping[str, Union[StrawberryType, type]]
+    ) -> Union[StrawberryType, type]:
+        return type_var_map[self.type_var.__name__]
+
+    @property
+    def is_graphql_generic(self) -> bool:
+        return True
+
+    def has_generic(self, type_var: TypeVar) -> bool:
+        return self.type_var == type_var
+
+    @property
+    def type_params(self) -> List[TypeVar]:
+        return [self.type_var]
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, StrawberryTypeVar):
+            return self.type_var == other.type_var
+        if isinstance(other, TypeVar):
+            return self.type_var == other
+
+        return super().__eq__(other)
+
+    def __hash__(self) -> int:
+        return hash(self.type_var)
+
+
+class WithStrawberryObjectDefinition(Protocol):
+    __strawberry_definition__: ClassVar[StrawberryObjectDefinition]
+
+
+def has_object_definition(
+    obj: Any,
+) -> TypeGuard[Type[WithStrawberryObjectDefinition]]:
+    if hasattr(obj, "__strawberry_definition__"):
+        return True
+    # TODO: Generics remove dunder members here, so we inject it here.
+    #  Would be better to avoid it somehow.
+    # https://github.com/python/cpython/blob/3a314f7c3df0dd7c37da7d12b827f169ee60e1ea/Lib/typing.py#L1152
+    if is_concrete_generic(obj):
+        concrete = obj.__origin__
+        if hasattr(concrete, "__strawberry_definition__"):
+            obj.__strawberry_definition__ = concrete.__strawberry_definition__
+            return True
+    return False
+
+
+@overload
+def get_object_definition(
+    obj: Any,
+    *,
+    strict: Literal[True],
+) -> StrawberryObjectDefinition: ...
+
+
+@overload
+def get_object_definition(
+    obj: Any,
+    *,
+    strict: bool = False,
+) -> Optional[StrawberryObjectDefinition]: ...
+
+
+def get_object_definition(
+    obj: Any,
+    *,
+    strict: bool = False,
+) -> Optional[StrawberryObjectDefinition]:
+    definition = obj.__strawberry_definition__ if has_object_definition(obj) else None
+    if strict and definition is None:
+        raise TypeError(f"{obj!r} does not have a StrawberryObjectDefinition")
+
+    return definition
 
 
 @dataclasses.dataclass(eq=False)
 class StrawberryObjectDefinition(StrawberryType):
```

Comment:
Do we want to keep `strawberry.types.types` or avoid it? In the issue I proposed renaming this file to `object_definition.py` because it originally only contained the `StrawberryObjectDefinition`. Now it contains a few classes more which we could instead move into separate files or into `__init__.py` itself. Just asking :)

Reply:
yes, I'd love to get rid of it 😊

Reply:
Uh, I like the `types.base` change 😊

Reply:
@DoctorJohn thanks for checking it! we did with @bellini666 :D